### DISCUSSION TITLE
Specification of `envmap.data` without an image file.

### DISCUSSION THREAD

**Discussion by maxfrei750** [mitsuba-renderer/mitsuba3#330] [Q&A] [answered]
I'd like to optimize the `data` attribute of an `envmap`. To do so, it needs to be initialized (e.g. with zeros or ones). I can do this using an image file and the `filename` attribute. However, IMHO it would be more elegant to do the initialization in memory, e.g. creating a suitable tensor while loading the scene. I tried to plug the initial data into the scene dictionary using a nested list, a numpy array or a drjit tensor:

e.g.
```
{
    ...
    "environment": {
        "type": "envmap",
        "data": dr.ones(dr.llvm.TensorXf, (8, 8, 4))
    }
    ...
}
```

However, I always receive errors like

```
RuntimeError("\u200b[xml_v.cpp:337] Unkown value type: <class 'drjit.llvm.TensorXf'>")
```

Is there a way to do this?

Thanks!

EDIT: Related: Should the range of the data be [0, 1]?

**Comment by Speierers** [comment, marked as answer]
Hi @maxfrei750 ,

Currently it isn't possible to pass a `TensorXf` object as property. I will add the possibility to directly pass a `mi.Bitmap` object (which can be created from a `TensorXf`) to the `envmap plugin.

**Reply by Speierers** [reply]
I added this feature in this commit: [9389c8](https://github.com/mitsuba-renderer/mitsuba3/commit/9389c8d1d16aa7a46d0a54f64eec1d10a1ae1ffd)

**Reply by maxfrei750** [reply]
Thanks for the quick implementation. It works really nicely. Here's a small snippet for others to get started:


```python
environment_size = 64

scene_dict = {
    "type": "scene",
    "environment": {
        "type":
            "envmap",
        "bitmap":
            mi.Bitmap(
                mi.TensorXf(mi.PCG32(size=environment_size**2).next_float32(),
                            shape=(environment_size, environment_size, 1))),
    }
}
```

It initializes the `envmap` with a random noise, which is preferable to a blank `envmap`, if I understood

https://github.com/mitsuba-renderer/mitsuba3/blob/ea513f73b65b8776afb75fdc8d40db4b1140345e/src/emitters/envmap.cpp#L173-L174 correctly.
